pkg/crypto/rings/cache.go

```diff
@@ -0,0 +1,201 @@
+// newRingFromPoints creates a new ring from points on the secp256k1 curve
+func newRingFromPoints(points []ringtypes.Point) (*ring.Ring, error) {
+	return ring.NewFixedKeyRingFromPublicKeys(ring_secp256k1.NewCurve(), points)
+}
+
+// getDelegatedPubKeysForAddress returns the ring used to sign a message for
+// the given application address, by querying the application module for it's
+// delegated pubkeys and converting them to points on the secp256k1 curve in
+// order to create the ring.
+func (rc *ringCache) getDelegatedPubKeysForAddress(
+	ctx context.Context,
+	appAddress string,
+) ([]ringtypes.Point, error) {
+	rc.ringPointsMu.Lock()
+	defer rc.ringPointsMu.Unlock()
+
+	// get the application's on chain state
+	req := &apptypes.QueryGetApplicationRequest{Address: appAddress}
+	res, err := rc.applicationQuerier.Application(ctx, req)
+	if err != nil {
+		return nil, ErrRingsAccountNotFound.Wrapf("address: %s [%v]", appAddress, err)
+	}
+
+	// create a slice of addresses for the ring
+	ringAddresses := make([]string, 0)
+	ringAddresses = append(ringAddresses, appAddress) // app address is index 0
+	if len(res.Application.DelegateeGatewayAddresses) < 1 {
+		// add app address twice to make the ring size of mininmum 2
+		// TODO_HACK: We are adding the appAddress twice because a ring
+		// signature requires AT LEAST two pubKeys. When the Application has
+		// not delegated to any gateways, we add the application's own address
+		// twice. This is a HACK and should be investigated as to what is the
+		// best approach to take in this situation.
+		ringAddresses = append(ringAddresses, appAddress)
+	} else if len(res.Application.DelegateeGatewayAddresses) > 0 {
+		// add the delegatee gateway addresses
+		ringAddresses = append(ringAddresses, res.Application.DelegateeGatewayAddresses...)
+	}
+
+	// get the points on the secp256k1 curve for the addresses
+	points, err := rc.addressesToPoints(ctx, ringAddresses)
+	if err != nil {
+		return nil, err
+	}
+
+	// update the cache overwriting the previous value
+	rc.ringPointsCache[appAddress] = points
+
+	// return the public key points on the secp256k1 curve
+	return points, nil
+}
+
+// addressesToPoints converts a slice of addresses to a slice of points on the
+// secp256k1 curve, by querying the account module for the public key for each
+// address and converting them to the corresponding points on the secp256k1 curve
+func (rc *ringCache) addressesToPoints(
+	ctx context.Context,
+	addresses []string,
+) ([]ringtypes.Point, error) {
+	curve := ring_secp256k1.NewCurve()
+	points := make([]ringtypes.Point, len(addresses))
+	for i, addr := range addresses {
+		pubKeyReq := &accounttypes.QueryAccountRequest{Address: addr}
+		pubKeyRes, err := rc.accountQuerier.Account(ctx, pubKeyReq)
+		if err != nil {
+			return nil, ErrRingsAccountNotFound.Wrapf("address: %s [%v]", addr, err)
+		}
+		var acc accounttypes.AccountI
+		reg := codectypes.NewInterfaceRegistry()
+		accounttypes.RegisterInterfaces(reg)
```

Comment:
OPTIONAL NIT: Decoupling business logic that does "codec and math" away from code that does "network interfacing" will really help readability, maintainability & testing.